modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java

```diff
@@ -210,27 +239,109 @@ private synchronized void releaseCachedClients() {
+     * <li>Supports gracefully shutting down the provider and the STS client.</li>
+     * </ul>
+     */
+    static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentialsProvider {
+
+        private STSAssumeRoleWithWebIdentitySessionCredentialsProvider credentialsProvider;
+        private AWSSecurityTokenService stsClient;
+
+        CustomWebIdentityTokenCredentialsProvider(Environment environment) {
+            // Check whether the original environment variable exists. If it doesn't,
+            // the system doesn't support AWS web identity tokens
+            if (System.getenv(AWS_WEB_IDENTITY_ENV_VAR) == null) {
+                return;
+            }
+            // Make sure that a readable symlink to the token file exists in the plugin config directory
+            Path webIdentityTokenFileSymlink = environment.configFile().resolve("repository-s3/aws-web-identity-token-file");
+            if (Files.exists(webIdentityTokenFileSymlink) == false) {
+                throw new IllegalStateException(
+                    "A Web Identity Token symlink in the config directory doesn't exist: " + webIdentityTokenFileSymlink
+                );
+            }
+            if (Files.isReadable(webIdentityTokenFileSymlink) == false) {
+                throw new IllegalStateException(
+                    "Unable to read a Web Identity Token symlink in the config directory: " + webIdentityTokenFileSymlink
+                );
+            }
+            String roleArn = System.getenv(AWS_ROLE_ARN_ENV_VAR);
+            String roleSessionName = System.getenv(AWS_ROLE_SESSION_NAME_ENV_VAR);
+            if (roleArn == null || roleSessionName == null) {
+                return;
+            }
+            AWSSecurityTokenServiceClientBuilder stsClientBuilder = AWSSecurityTokenServiceClient.builder();
+
+            // Just for testing
+            String customStsEndpoint = System.getProperty("com.amazonaws.sdk.stsMetadataServiceEndpointOverride");
+            if (customStsEndpoint != null) {
+                stsClientBuilder.withEndpointConfiguration(new AwsClientBuilder.EndpointConfiguration(customStsEndpoint, null));
+            }
+            stsClientBuilder.withCredentials(new AWSStaticCredentialsProvider(new AnonymousAWSCredentials()));
+            stsClient = SocketAccess.doPrivileged(stsClientBuilder::build);
+            credentialsProvider = new STSAssumeRoleWithWebIdentitySessionCredentialsProvider.Builder(
+                roleArn,
+                roleSessionName,
+                webIdentityTokenFileSymlink.toString()
+            ).withStsClient(stsClient).build();
+        }
+
+        boolean isActive() {
+            return credentialsProvider != null;
+        }
+
+        @Override
+        public AWSCredentials getCredentials() {
+            return credentialsProvider != null ? credentialsProvider.getCredentials() : new AnonymousAWSCredentials();
```

Comment:
It shouldn't because we don't call `new AWSCredentialsProviderChain(webIdentityTokenCredentialsProvider, new EC2ContainerCredentialsProviderWrapper())` unless the credential provider is not null. Let me throw an exception in case it get called when credentialsProvider is null.